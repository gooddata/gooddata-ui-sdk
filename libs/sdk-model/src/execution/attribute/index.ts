// (C) 2019-2023 GoodData Corporation
import { Identifier, isUriRef, ObjRef, isIdentifierRef } from "../../objRef/index.js";
import isEmpty from "lodash/isEmpty.js";
import { invariant } from "ts-invariant";

/**
 * Attribute is our nomenclature for 'dimension' as typically used in multi-dimensional BI modeling.
 *
 * @remarks
 * The attribute specifies on what dimension to slice/dice the measures.
 *
 * In GoodData, each attribute can have multiple different display forms. Display forms essentially provide different
 * representations for elements in the dimension.
 *
 * @public
 */
export interface IAttributeBody {
    /**
     * User-assigned id of the attribute. The value is used to cross-reference model elements that are part
     * of an execution.
     *
     * The constraint is that the local identifiers of all measures and attributes in an execution definition MUST
     * be unique.
     */
    localIdentifier: Identifier;

    /**
     * Reference to display form to use for element representations.
     *
     * The display form can be referenced by either URL of the display form resource on backend OR by
     * unique, backend-recognized, identifier of the display form.
     *
     * Note: specifying display forms by URI is discouraged and WILL be deprecated in the future. The
     * specification of URL has several drawbacks:
     *
     * -  Display form URLs are tied to particular analytical workspace; this makes any application that uses
     *    URL-specified display forms workspace-specific.
     * -  Display form URLs are not supported by all GoodData backends == this makes any application that uses
     *    URL-specified display form backend-specific
     */
    displayForm: ObjRef;

    /**
     * This is user-assigned alias of the attribute; it is opaque for the backend. When backend returns
     * metadata about execution results, it WILL include this user-assigned alias in the metadata.
     */
    alias?: string;

    /**
     * Indicates whether to show all values of given attribute even if the data bound to those values
     * are not available
     */
    showAllValues?: boolean;
}

/**
 * Attribute is our nomenclature for 'dimension' as typically used in multi-dimensional BI modeling.
 *
 * @remarks
 * The attribute specifies on what dimension to slice/dice the measures.
 *
 * In GoodData, each attribute can have multiple different display forms. Display forms essentially provide different
 * representations for elements in the dimension.
 *
 * @public
 */
export interface IAttribute {
    attribute: IAttributeBody;
}

/**
 * Type guard checking whether object is an instance of IAttribute.
 *
 * @public
 */
export function isAttribute(obj: unknown): obj is IAttribute {
    return !isEmpty(obj) && (obj as IAttribute).attribute !== undefined;
}

/**
 * Defines function signature for measure predicates.
 *
 * @public
 */
export type AttributePredicate = (attribute: IAttribute) => boolean;

/**
 * Predicate that returns true for any attribute.
 *
 * @public
 */
export const anyAttribute: AttributePredicate = (_) => true;

/**
 * Factory function for attribute predicate which evaluates true for attributes that match particular ID.
 *
 * @public
 */
export const idMatchAttribute: (id: string) => AttributePredicate = (id) => (attr) =>
    attr.attribute.localIdentifier === id;

//
//
//

/**
 * Gets local identifier of an attribute.
 *
 * @remarks
 * For convenience and fluency, this function accepts both attribute object and identifier.
 *
 * @param attributeOrId - attribute to work with or the identifier
 * @returns value of local identifier
 * @public
 */
export function attributeLocalId(attributeOrId: IAttribute | Identifier): string {
    invariant(attributeOrId, "attribute must be specified");

    return typeof attributeOrId === "string" ? attributeOrId : attributeOrId.attribute.localIdentifier;
}

/**
 * Gets URI of attribute's display form to use and get attribute element values from.
 *
 * @param attribute - attribute to work with, may be undefined == result is undefined
 * @returns display form URI as string, undefined if display form not specified using URI
 * @public
 */
export function attributeUri(attribute: IAttribute): string | undefined {
    invariant(attribute, "attribute must be specified");

    return isUriRef(attribute.attribute.displayForm) ? attribute.attribute.displayForm.uri : undefined;
}

/**
 * Gets identifier of attribute's display form to use and get attribute element values from.
 *
 * @param attribute - attribute to work with, may be undefined == result is undefined
 * @returns display form identifier as string, undefined if display for not specified using identifier
 * @public
 */
export function attributeIdentifier(attribute: IAttribute): string | undefined {
    invariant(attribute, "attribute must be specified");

    return isIdentifierRef(attribute.attribute.displayForm)
        ? attribute.attribute.displayForm.identifier
        : undefined;
}

/**
 * Gets an attribute alias.
 *
 * @param attribute - attribute to work with
 * @returns value of attribute alias
 * @public
 */
export function attributeAlias(attribute: IAttribute): string | undefined {
    invariant(attribute, "attribute must be specified");

    return attribute.attribute.alias;
}

/**
 * Gets an attribute show all values property.
 *
 * @param attribute - attribute to work with
 * @returns value of attribute show all values property
 * @public
 */
export function attributeShowAllValues(attribute: IAttribute): boolean | undefined {
    invariant(attribute, "attribute must be specified");

    return attribute.attribute.showAllValues;
}

/**
 * Gets an attribute display form object ref.
 *
 * @param attribute - attribute to work with
 * @returns value of attribute display form object ref
 * @public
 */
export function attributeDisplayFormRef(attribute: IAttribute): ObjRef {
    invariant(attribute, "attribute must be specified");

    return attribute.attribute.displayForm;
}

/**
 * Given list of attributes, returns first-found attribute matching the provided predicate.
 *
 * @remarks
 * If no predicate is provided, then the function defaults to anyAttribute predicate - meaning first found attribute
 * will be returned.
 *
 * This function also provides convenience to find attribute by its local identifier - if you pass predicate as
 * string the function will automatically create idMatchAttribute predicate.
 *
 * @param attributes - list of attributes to work with, must be specified
 * @param idOrFun - attribute identifier or instance of AttributePredicate; {@link anyAttribute} predicate is default
 * @public
 */
export function attributesFind(
    attributes: IAttribute[],
    idOrFun: string | AttributePredicate = anyAttribute,
): IAttribute | undefined {
    invariant(attributes, "attributes must be specified");

    if (!attributes.length) {
        return;
    }

    const predicate = typeof idOrFun === "string" ? idMatchAttribute(idOrFun) : idOrFun;

    return attributes.find(predicate);
}
