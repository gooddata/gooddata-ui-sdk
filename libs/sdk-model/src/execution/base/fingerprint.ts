// (C) 2019-2020 GoodData Corporation

import stringify from "json-stable-stringify";
import merge from "lodash/merge.js";
import { IDimension } from "./dimension.js";

type DimensionPropsToDefault = Pick<IDimension, "totals">;

/**
 * Calculates dimension fingerprint; ensures that the optional vs default values are correctly reflected in
 * the fingerprint.
 *
 * @internal
 */
export function dimensionFingerprint(dim: IDimension): string {
    const dimDefaults: DimensionPropsToDefault = {
        totals: [],
    };

    const withDefaultTotals: IDimension = merge(dimDefaults, dim);

    return stringify(withDefaultTotals);
}
