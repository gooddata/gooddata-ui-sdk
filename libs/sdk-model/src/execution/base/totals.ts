// (C) 2019-2022 GoodData Corporation

import isEmpty from "lodash/isEmpty.js";
import { IMeasure, measureLocalId } from "../measure/index.js";
import { attributeLocalId, IAttribute } from "../attribute/index.js";
import { invariant } from "ts-invariant";
import { Identifier } from "../../objRef/index.js";

/**
 * Supported types of totals.
 *
 * @public
 */
export type TotalType = "sum" | "avg" | "max" | "min" | "med" | "nat";

/**
 * Describes type and granularity for calculation of Totals.
 *
 * @remarks
 * Total is calculated for particular measure and on some granularity - specified by an attribute
 * by which the measure is sliced by.
 *
 * @public
 */
export interface ITotal {
    /**
     * Type of total calculation.
     */
    type: TotalType;

    /**
     * Local identifier of measure for which to calculate total.
     */
    measureIdentifier: Identifier;

    /**
     * Local identifier of attribute - specifies granularity of the calculation.
     */
    attributeIdentifier: Identifier;

    /**
     * Specifies custom name for the calculated total. This will be included in result metadata.
     */
    alias?: string;
}

//
// Type guards
//

/**
 * Type-guard checking whether an object is a Total.
 *
 * @public
 */
export function isTotal(obj: unknown): obj is ITotal {
    return (
        !isEmpty(obj) &&
        (obj as ITotal).type !== undefined &&
        (obj as ITotal).measureIdentifier !== undefined &&
        (obj as ITotal).attributeIdentifier !== undefined
    );
}

//
//
//

/**
 * Creates a new total.
 *
 * @param type - type of total, one of the enumerated types
 * @param measureOrId - measure instance OR measure local identifier
 * @param attributeOrId - attribute instance OR attribute local identifier
 * @param alias - provide custom name (alias) for the total; this will be included in the computed results
 * @returns new total
 * @public
 */
export function newTotal(
    type: TotalType,
    measureOrId: IMeasure | Identifier,
    attributeOrId: IAttribute | Identifier,
    alias?: string,
): ITotal {
    invariant(type, "total type must be specified");
    invariant(measureOrId, "measure or measure local id must be specified");
    invariant(attributeOrId, "attribute or attribute local id must be specified");

    const measureIdentifier = measureLocalId(measureOrId);
    const attributeIdentifier = attributeLocalId(attributeOrId);
    const aliasProp = alias ? { alias } : {};

    return {
        type,
        measureIdentifier,
        attributeIdentifier,
        ...aliasProp,
    };
}

/**
 * Tests whether total instance represents a native total = a roll-up total.
 *
 * @public
 */
export function totalIsNative(total: ITotal): boolean {
    invariant(total, "total must be specified");

    return total.type === "nat";
}
