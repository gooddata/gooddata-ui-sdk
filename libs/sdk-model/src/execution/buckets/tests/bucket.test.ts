// (C) 2019-2020 GoodData Corporation

import {
    applyRatioRule,
    IAttributeOrMeasure,
    bucketAttributeIndex,
    bucketAttribute,
    bucketAttributes,
    bucketIsEmpty,
    bucketItems,
    bucketMeasureIndex,
    bucketMeasure,
    bucketMeasures,
    bucketTotals,
    ComputeRatioRule,
    newBucket,
    BucketItemModifications,
    IBucket,
    bucketModifyItems,
} from "../index";
import { Account, Activity, Velocity, Won } from "../../../../__mocks__/model";
import { InvariantError } from "ts-invariant";
import { ITotal, newTotal } from "../../base/totals";
import { attributeLocalId, IAttribute, idMatchAttribute } from "../../attribute";
import { idMatchMeasure, IMeasure, measureLocalId, isMeasure } from "../../measure";
import { modifySimpleMeasure } from "../../../index";

describe("newBucket", () => {
    const Scenarios: Array<[string, any, any[]]> = [
        ["empty bucket", "emptyBucket", [undefined]],
        ["bucket with single measure", "measureBucket", [Won]],
        ["bucket with multiple measures", "measureBucket", [Won, Velocity.Sum]],
        ["bucket with single attribute", "attributeBucket", [Account.Default]],
        ["bucket with multiple attributes", "attributeBucket", [Account.Default, Account.Name]],
        [
            "bucket with mix of attributes and measures",
            "mixedBucket",
            [Won, Account.Name, Velocity.Sum, Account.Default],
        ],
        ["bucket with totals", "totalsBucket", [Won, Account.Name, newTotal("sum", Won, Account.Name)]],
        ["bucket with totals when items not in bucket", "totalsBucket", [newTotal("sum", Won, Account.Name)]],
    ];

    it.each(Scenarios)("should create %s", (_desc, localIdArg, contentArg) => {
        expect(newBucket(localIdArg, ...contentArg)).toMatchSnapshot();
    });

    it("should throw if local identifier not specified", () => {
        expect(() => newBucket(undefined as any, Won, Account.Name)).toThrowError(InvariantError);
    });

    it("should throw if content is invalid type", () => {
        expect(() => newBucket("invalidBucket", Won, {} as any, Account.Name)).toThrowError(InvariantError);
    });
});

const Total = newTotal("sum", Won, Account.Name);
const EmptyBucket = newBucket("EmptyBucket");
const BucketWithOnlyTotals = newBucket("BucketWithOnlyTotals", newTotal("sum", Won, Account.Name));
const BucketWithSingleMeasure = newBucket("BucketWithSingleMeasure", Won);
const BucketWithManyMeasures = newBucket("BucketWithManyMeasures", Won, Velocity.Sum, Velocity.Max);
const BucketWithSingleAttr = newBucket("BucketWithSingleAttr", Account.Name);
const BucketWithManyAttrs = newBucket("BucketWithManyAttrs", Account.Name, Account.Default, Activity.Subject);
const BucketWithMeasureAndAttr = newBucket("BucketWithMeasureAndAttr", Won, Account.Name);
const BucketWithEverything = newBucket("BucketWithMeasureAndAttr", Won, Account.Name, Total);

const InvalidScenarios: Array<[string, any]> = [
    ["undefined input", undefined],
    ["null input", null],
];

describe("bucketIsEmpty", () => {
    const Scenarios: Array<[boolean, string, any]> = [
        [true, "empty bucket", EmptyBucket],
        [false, "bucket with only totals", BucketWithOnlyTotals],
        [false, "bucket with single measure", BucketWithSingleMeasure],
        [false, "bucket with single attr", BucketWithSingleAttr],
        [false, "bucket with all possible inputs", BucketWithEverything],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, input) => {
        expect(bucketIsEmpty(input)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketIsEmpty(input)).toThrow();
    });
});

describe("bucketAttributeIndex", () => {
    const Scenarios: Array<[string, any, any, number]> = [
        ["no attribute in empty bucket", EmptyBucket, undefined, -1],
        ["no attribute in measure-only bucket", BucketWithManyMeasures, undefined, -1],
        ["no attribute if localId empty", BucketWithManyAttrs, "", -1],
        ["no attribute if localId matches nothing", BucketWithManyAttrs, "noSuchLocalId", -1],
        ["first attribute when no predicate provided", BucketWithManyAttrs, undefined, 0],
        ["first attribute in mixed attr & measure bucket", BucketWithMeasureAndAttr, undefined, 1],
        [
            "attribute by local id if predicate is a string",
            BucketWithManyAttrs,
            attributeLocalId(Account.Default),
            1,
        ],
        ["no attribute if no predicate match", BucketWithManyAttrs, () => false, -1],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketAttributeIndex(bucketArg, predicateArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketAttributeIndex(input)).toThrow();
    });
});

describe("bucketAttribute", () => {
    const Scenarios: Array<[string, any, any, IAttribute | undefined]> = [
        ["no attribute in empty bucket", EmptyBucket, undefined, undefined],
        ["no attribute in measure-only bucket", BucketWithManyMeasures, undefined, undefined],
        ["no attribute if localId empty", BucketWithManyAttrs, "", undefined],
        ["no attribute if localId matches nothing", BucketWithManyAttrs, "noSuchLocalId", undefined],
        ["first attribute when no predicate provided", BucketWithManyAttrs, undefined, Account.Name],
        ["first attribute in mixed attr & measure bucket", BucketWithMeasureAndAttr, undefined, Account.Name],
        [
            "attribute by local id if predicate is a string",
            BucketWithManyAttrs,
            attributeLocalId(Account.Default),
            Account.Default,
        ],
        ["no attribute if no predicate match", BucketWithManyAttrs, () => false, undefined],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketAttribute(bucketArg, predicateArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketAttribute(input)).toThrow();
    });
});

describe("bucketAttributes", () => {
    const Scenarios: Array<[string, any, any, IAttribute[]]> = [
        ["no attributes in empty bucket", EmptyBucket, undefined, []],
        ["no attributes in measure-only bucket", BucketWithManyMeasures, undefined, []],
        ["no attributes in attr bucket but no predicate match", BucketWithManyAttrs, () => false, []],
        ["attributes in single attr bucket", BucketWithSingleAttr, undefined, [Account.Name]],
        [
            "attributes in many attr bucket",
            BucketWithManyAttrs,
            undefined,
            [Account.Name, Account.Default, Activity.Subject],
        ],
        [
            "single attribute matching predicate",
            BucketWithManyAttrs,
            idMatchAttribute(attributeLocalId(Account.Name)),
            [Account.Name],
        ],
        ["attributes in mixed bucket", BucketWithMeasureAndAttr, undefined, [Account.Name]],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketAttributes(bucketArg, predicateArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketAttributes(input)).toThrow();
    });
});

describe("bucketMeasureIndex", () => {
    const Scenarios: Array<[string, any, any, number]> = [
        ["no measure in empty bucket", EmptyBucket, undefined, -1],
        ["no measure in attr-only bucket", BucketWithManyAttrs, undefined, -1],
        ["no measure if localId empty", BucketWithManyMeasures, "", -1],
        ["no measure if localId matches nothing", BucketWithManyMeasures, "noSuchLocalId", -1],
        ["first measure when no predicate provided", BucketWithManyMeasures, undefined, 0],
        ["first measure in mixed attr & measure bucket", BucketWithMeasureAndAttr, undefined, 0],
        [
            "measure by local id if predicate is a string",
            BucketWithManyMeasures,
            measureLocalId(Velocity.Sum),
            1,
        ],
        ["no measure if no predicate match", BucketWithManyMeasures, () => false, -1],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketMeasureIndex(bucketArg, predicateArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketMeasureIndex(input)).toThrow();
    });
});

describe("bucketMeasure", () => {
    const Scenarios: Array<[string, any, any, IMeasure | undefined]> = [
        ["no measure in empty bucket", EmptyBucket, undefined, undefined],
        ["no measure in attr-only bucket", BucketWithManyAttrs, undefined, undefined],
        ["no measure if localId empty", BucketWithManyMeasures, "", undefined],
        ["no measure if localId matches nothing", BucketWithManyMeasures, "noSuchLocalId", undefined],
        ["first measure when no predicate provided", BucketWithManyMeasures, undefined, Won],
        ["first measure in mixed attr & measure bucket", BucketWithMeasureAndAttr, undefined, Won],
        [
            "measure by local id if predicate is a string",
            BucketWithManyMeasures,
            measureLocalId(Velocity.Sum),
            Velocity.Sum,
        ],
        ["no measure if no predicate match", BucketWithManyMeasures, () => false, undefined],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketMeasure(bucketArg, predicateArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketMeasure(input)).toThrow();
    });
});

describe("bucketMeasures", () => {
    const Scenarios: Array<[string, any, any, IMeasure[]]> = [
        ["no measures in empty bucket", EmptyBucket, undefined, []],
        ["no measures in attr-only bucket", BucketWithManyAttrs, undefined, []],
        ["no measures in measure bucket but no predicate match", BucketWithManyMeasures, () => false, []],
        ["measures in single measure bucket", BucketWithSingleMeasure, undefined, [Won]],
        [
            "measures in many measure bucket",
            BucketWithManyMeasures,
            undefined,
            [Won, Velocity.Sum, Velocity.Max],
        ],
        [
            "single measure matching predicate",
            BucketWithManyMeasures,
            idMatchMeasure(measureLocalId(Velocity.Sum)),
            [Velocity.Sum],
        ],
        ["measures in mixed bucket", BucketWithMeasureAndAttr, undefined, [Won]],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketMeasures(bucketArg, predicateArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketMeasures(input)).toThrow();
    });
});

describe("bucketItems", () => {
    const Scenarios: Array<[string, any, IAttributeOrMeasure[]]> = [
        ["no items in empty bucket", EmptyBucket, []],
        ["items in populated bucket", BucketWithMeasureAndAttr, [Won, Account.Name]],
    ];

    it.each(Scenarios)("should return %s", (_desc, bucketArg, expectedResult) => {
        expect(bucketItems(bucketArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketItems(input)).toThrow();
    });
});

describe("bucketTotals", () => {
    const Scenarios: Array<[string, any, ITotal[]]> = [
        ["no totals in empty bucket", EmptyBucket, []],
        ["no totals in bucket without totals", BucketWithMeasureAndAttr, []],
        ["totals in populated bucket", BucketWithEverything, [Total]],
    ];

    it.each(Scenarios)("should return %s", (_desc, bucketArg, expectedResult) => {
        expect(bucketTotals(bucketArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketTotals(input)).toThrow();
    });
});

describe("applyRatioRule", () => {
    const MeasureWithRatio1 = modifySimpleMeasure(Won, (m) => m.ratio());
    const MeasureWithRatio2 = modifySimpleMeasure(Velocity.Avg, (m) => m.ratio());

    const Scenarios: Array<[string, any, any, any]> = [
        [
            "apply SINGLE rule by default",
            [MeasureWithRatio1],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [MeasureWithRatio1],
        ],
        [
            "keep all ratios if rule is ALWAYS",
            [MeasureWithRatio1, MeasureWithRatio2],
            ComputeRatioRule.ANY_MEASURE,
            [MeasureWithRatio1, MeasureWithRatio2],
        ],
        [
            "reset all ratios if rule is NEVER",
            [MeasureWithRatio1, MeasureWithRatio2],
            ComputeRatioRule.NEVER,
            [Won, Velocity.Avg],
        ],
        [
            "reset all ratios if rule is SINGLE and there are two ratio measures",
            [MeasureWithRatio1, MeasureWithRatio2],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [Won, Velocity.Avg],
        ],
        [
            "keep ratio of single ratio measure",
            [MeasureWithRatio1],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [MeasureWithRatio1],
        ],
        [
            "keep ratio of single ratio measure when mixed with attributes",
            [MeasureWithRatio1, Account.Name],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [MeasureWithRatio1, Account.Name],
        ],
        [
            "retain attributes as-is when SINGLE rule",
            [Activity.Subject, MeasureWithRatio1, Account.Name],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [Activity.Subject, MeasureWithRatio1, Account.Name],
        ],
        [
            "retain attributes as-is when NEVER rule",
            [Activity.Subject, MeasureWithRatio1, Account.Name],
            ComputeRatioRule.NEVER,
            [Activity.Subject, Won, Account.Name],
        ],
    ];

    it.each(Scenarios)("should %s", (_desc, itemsArg, ruleArg, expectedResult) => {
        expect(applyRatioRule(itemsArg, ruleArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => bucketTotals(input)).toThrow();
    });
});

describe("bucketModifyItems", () => {
    const modifications: BucketItemModifications = (bucketItem: IAttributeOrMeasure): IAttributeOrMeasure => {
        if (isMeasure(bucketItem)) {
            (bucketItem as IMeasure).measure.title = "Modified measure title";
        } else {
            (bucketItem as IAttribute).attribute.alias = "Modified attribute alias";
        }
        return bucketItem;
    };

    it("should return another bucket with no bucket item", () => {
        const modifiedBucket: IBucket = bucketModifyItems(EmptyBucket, modifications);
        expect(modifiedBucket !== EmptyBucket).toBe(true);
        expect(modifiedBucket.items.length).toBe(0);
    });

    it("should return another bucket with modified bucket items", () => {
        const modifiedBucket: IBucket = bucketModifyItems(BucketWithMeasureAndAttr, modifications);
        expect(modifiedBucket !== BucketWithMeasureAndAttr).toBe(true);
        expect(modifiedBucket.items.length).toBe(2);
        expect((modifiedBucket.items[0] as IMeasure).measure.title).toBe("Modified measure title");
        expect((modifiedBucket.items[1] as IAttribute).attribute.alias).toBe("Modified attribute alias");
    });
});
