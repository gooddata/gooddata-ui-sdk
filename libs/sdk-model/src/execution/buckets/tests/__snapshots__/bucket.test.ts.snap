// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`newBucket > should create bucket with mix of attributes and measures 1`] = `
{
  "items": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.account.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.account.id.name",
      },
    },
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "aggregation": "sum",
            "item": {
              "identifier": "fact.stagehistory.velocity",
            },
          },
        },
        "localIdentifier": "m_fact.stagehistory.velocity_sum",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.account.id",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.account.id",
      },
    },
  ],
  "localIdentifier": "mixedBucket",
}
`;

exports[`newBucket > should create bucket with multiple attributes 1`] = `
{
  "items": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.account.id",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.account.id",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.account.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.account.id.name",
      },
    },
  ],
  "localIdentifier": "attributeBucket",
}
`;

exports[`newBucket > should create bucket with multiple measures 1`] = `
{
  "items": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "aggregation": "sum",
            "item": {
              "identifier": "fact.stagehistory.velocity",
            },
          },
        },
        "localIdentifier": "m_fact.stagehistory.velocity_sum",
      },
    },
  ],
  "localIdentifier": "measureBucket",
}
`;

exports[`newBucket > should create bucket with single attribute 1`] = `
{
  "items": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.account.id",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.account.id",
      },
    },
  ],
  "localIdentifier": "attributeBucket",
}
`;

exports[`newBucket > should create bucket with single measure 1`] = `
{
  "items": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
  ],
  "localIdentifier": "measureBucket",
}
`;

exports[`newBucket > should create bucket with totals 1`] = `
{
  "items": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.account.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.account.id.name",
      },
    },
  ],
  "localIdentifier": "totalsBucket",
  "totals": [
    {
      "attributeIdentifier": "a_label.account.id.name",
      "measureIdentifier": "m_afSEwRwdbMeQ",
      "type": "sum",
    },
  ],
}
`;

exports[`newBucket > should create bucket with totals when items not in bucket 1`] = `
{
  "items": [],
  "localIdentifier": "totalsBucket",
  "totals": [
    {
      "attributeIdentifier": "a_label.account.id.name",
      "measureIdentifier": "m_afSEwRwdbMeQ",
      "type": "sum",
    },
  ],
}
`;

exports[`newBucket > should create empty bucket 1`] = `
{
  "items": [],
  "localIdentifier": "emptyBucket",
}
`;

exports[`newBucket > should throw error if attribute has more than one display forms 1`] = `[Invariant Violation: Contents of a bucket must be either attribute, measure or total. It looks like you used an attribute from generated metadata containing more than one display form. Use one of the following display forms instead: Subject: label.activity.id.subject, Default: label.activity.id.]`;

exports[`newBucket > should throw error if content invalid object containing valid aggregations 1`] = `[Invariant Violation: Contents of a bucket must be either attribute, measure or total. It looks like you used an object fact.stagehistory.duration from generated metadata. You need to use one of the following aggregation functions instead: Sum, Count, Avg, Min, Max, Median, Runsum.]`;

exports[`newBucket > should throw if content not an object 1`] = `[Invariant Violation: Contents of a bucket must be either attribute, measure or total. Got unsupported content of type string: string.]`;

exports[`newBucket > should throw if unknown object 1`] = `[Invariant Violation: Contents of a bucket must be either attribute, measure or total. Got unknown content object: {}.]`;
